### ISSUE TITLE
[Suggestion] templateflow integration for atlases

### ISSUE THREAD

**Issue by wiheto** [nbraingroup/fmridenoise#29]
2 observations: 

1. Only Schaeffer 200 is included with fmridenoise.
2. It is currently in FSLMNI152 space whereas default fMRIprep output is 2009c. 

Suggestions: 

1. input argument of desired space to be working with (default being 2009c as thats the fmriprep default). But allow people to put in whatever space they want from fMRIprep. 
2. Atlases get pulled from the space they are in from templateflow. 

Positives: 

1. Ensures correct space is used with fmriprep output. 
2. Requires no parcellation maintenance from fmridenoise. 
3. Also avoids any misuse of applying an MNI atlas to output that is only in subject space 

Negatives: 

1. TemplateFlow is still in development, but for current purposes (default fMRIprep output and Scaeffer atlas) it will expand on current functionality. 

If suggestion is liked, I can try and implement this. Changes that will be made will only revolve around when the parcellation file is called. May take a week or two before I get round to this though.

**Comment by kbonna** [comment]
That was exactly our plan for future development. We wanted to add more parcellations and allow user to choose among them. Users could also add their own parcellations, but this require some additional instructions for them to follow specification. Your second observation is a great point, we didn't notice that before. I actually fixed bids query to look only for 2009c template. I will add optional space argument and set 2009c as a default (this is quick fix).

If you want to help us implementing this functionality, that would be really great! Fell free to ask any questions.

**Comment by kfinc** [comment]
Great suggestion @wiheto ! 

I'm not familiar with TemplateFlow, but it would be much better to pull parcellations in desired space from tools that already exist than storing them in fmridenoise. How TemplateFlow adjusts parcellations in other MNI spaces to match them with the default fMRIPrep's 2009c? 

It would be great if you decide to help with this enhancement!

**Comment by wiheto** [comment]
TemplateFlow is something that is now integrated with fMRIprep: https://github.com/templateflow/templateflow

Basically, it is a collection of templates across multiple different spaces. And different atlases across the different spaces will get added over time (something I'm partially working on). Schaeffer is already there. It also has an API which pulls the atlas to the user's computer if required that is really easy to use. Example:  

```
import templateflow.api as tf
import nibabel as nib
atlaspathj = tf.get('MNI152NLin2009cAsym', atlas='Schaefer2018', desc='200Parcels7Networks', resolution=1)
img = nib.load(str(atlaspath))
```

This will both download the image if its not found on your computer. 

In this example: `MNI152NLin2009cAsym` could be default for a space flag. However, this can be  obtained automatically from pybids as its values of the `space` key. If a user has asked for a different output space from fMRIprep, then it can try and pull that atlas from templateflow (while the coverage of all spaces in templateflow is not perfect yet, this will expand over time and means you won't have to worry about that part). 

Then just a flag for atlas (default: Schaefer2018) and atlas_desc (default: 200Parcels7Networks)

**Comment by wiheto** [comment]
Hopefully that small background about templateflow made sense. i'll have a look in a week or two how easy it will be to add it to fMRIdenoise as I test it some more with a dataset I'm working with

**Comment by kfinc** [comment]
Thanks @wiheto ! 

I tried to load Schaefer2018 atlas with the code you provided, but I've got an empty array in `atlaspath`. What could go wrong?

Does TemplateFlow provide atlas metadata i.e. parcels coordinates? In our case, we also need coordinates to calculate the distance matrix.

**Comment by wiheto** [comment]
Weird. That code works for me. Do you have version 0.4.1?

> Does TemplateFlow provide atlas metadata i.e. parcels coordinates? In our case, we also need coordinates to calculate the distance matrix. 

There is metadata that you can pull down as well (just change). An example of the meta-data is here "https://github.com/templateflow/tpl-MNI152NLin2009cAsym/blob/master/tpl-MNI152NLin2009cAsym_atlas-Schaefer2018_desc-600Parcels17Networks_dseg.tsv" which does not have the coordinates. What kind of coordinates do you mean? Like the center of the parcel?

**Comment by kfinc** [comment]
Hi @wiheto! Sorry for missing your last question. 
In terms of coordinates, we need centroid coordinates for each parcel, as provided here: https://github.com/ThomasYeoLab/CBIG/tree/master/stable_projects/brain_parcellation/Schaefer2018_LocalGlobal/Parcellations/MNI/Centroid_coordinates.

I'm still not able to load any template with `templateflow`. I'm getting an empty list for `atlaspath`. I have a version 0.1.8 now. Any idea why that happens?

**Comment by wiheto** [comment]
So the centroid conversion is something that is possible and I will know how to do soon. But I'll hold it off for now (although it is something that could be calculated on the fly as well). 

I had that problem the first time I installed it and then upgrade to a later templateflow version. I *think* the solution was to delete ~/.cache/templateflow/ and try again.

**Comment by manzouri** [comment]
> Hopefully that small background about templateflow made sense. i'll have a look in a week or two how easy it will be to add it to fMRIdenoise as I test it some more with a dataset I'm working with

Hi , I wonder if it possible to specify different space in fmridenoise? Apparently the default is "MNI152NLin2009cAsym " but I have done fmriprep with "MNI152NLin6Asym", when I try to run fmridenoise I get error!

**Comment by nicolasdarmanthe** [comment]
> > Hopefully that small background about templateflow made sense. i'll have a look in a week or two how easy it will be to add it to fMRIdenoise as I test it some more with a dataset I'm working with
> 
> Hi , I wonder if it possible to specify different space in fmridenoise? Apparently the default is "MNI152NLin2009cAsym " but I have done fmriprep with "MNI152NLin6Asym", when I try to run fmridenoise I get error!

I don't think it is currently possible to use a different template space (though I imagine it wouldn't be too hard to implement). fmridenoise looks quite promising but unfortunately it's still pretty immature. The docs are minimal, and you have to go digging pretty deep through the source code to understand what is going on behind the scenes.

fmridenoise also relies on nipype, which doesn't work out of the box on windows. You can turnfmridenoise into a docker image to get around this, but again this is not straight forward. I think there is still a lot of work to be done before fmridenoise can be used for widely.
